### PR TITLE
Ask Mode request: Fix IndexOutOfRangeException when running operation actions over Roslyn sources

### PR DESCRIPTION
**Pull request by JohnHamby**
Running operation actions over Roslyn sources was producing an IndexOutOfRangeException when attempting to process the arguments to some constructions of DiagnosticDescriptor.

The root issue requires these circumstances:
  1) A target method or constructor with optional parameters followed by a params parameter, e.g.:
        void foo(int a, int b = 10, int c = 10, params int[] d);
  2) A call that specifies some of the optional parameters by name and does not explicitly specify an argument matching the params parameter, e.g.:
        foo(1, c: 2);

@genlu @mavasani @ManishJayaswal @Pilchie might care. This is an Ask Mode request.

### CODE REVIEW

**Comment by JohnHamby** [comment]
@srivatsn might also care.

**Comment by genlu** [comment]
For ask mode change, please rebase to target stabilization branch instead.

**Comment by genlu** [comment]
I don't know what was the exact cause of crashes reported in: #8084, #8748 and #8841. Tag those here since it seems related.

**Comment by mavasani** [comment]
@genlu I think all of these are the same issue.

**Comment by mavasani** [comment]
:+1:, but you may want a signoff from @genlu

**Comment by JohnHamby** [comment]
If the change is approved for Ask Mode, or if it's a prerequisite for approval, I may need some help creating an appropriately rebased PR for the stabilization branch.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs [hunk 1]
```diff
@@ -1145,6 +1145,27 @@ public void M1()
         M0(1, new int[] { 2, 3, 4 });
         M0(1, new int[] { 2, 3, 4, 5 });
         M0(1, new int[] { 2, 3, 4, 5, 6 });
+        M2(1, c: 2);
+        D d = new D(3, c: 40);
+        d = new D(""Hello"", 1, 2, 3, 4);
+        d = new D(""Hello"", new int[] { 1, 2, 3, 4 });
+        d = new D(""Hello"", 1, 2, 3);
+        d = new D(""Hello"", new int[] { 1, 2, 3 });
```
Comment (genlu) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/OperationAnalyzerTests.cs]: What happens if we do something like this?
```cs
var array = new int[] { 1, 2, 3, 4 }
d = new D("Hello", array );
```

Will `array` be treated as a params array argument?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 2]
```diff
@@ -97,22 +97,22 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
...
+                        // The parameter is optional with a default value.
                         arguments.Add(new Argument(ArgumentKind.DefaultValue, parameter, new Literal(parameter.ExplicitDefaultConstantValue, parameter.Type, null)));
                     }
                     else
                     {
-                        // If corresponding parameter is Param array, then this means 0 element is provided and an Argument of kind == ParamArray will be added, 
-                        // otherwise it is an error and null is added.
-                        arguments.Add(DeriveArgument(parameterIndex, argumentIndex, boundArguments, argumentNames, argumentRefKinds, parameters, invocationSyntax));
+                        // The parameter might be a params array, in which case an empty array will be provided.
+                        arguments.Add(DeriveArgument(parameterIndex, boundArguments.Length, boundArguments, argumentNames, argumentRefKinds, parameters, invocationSyntax));
```
Comment (genlu) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: It seems we would reach this line only the parameter is a params array and no value is provided, is that right?

Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: That is true if the invocation is semantically valid. It's possible to get here if an argument is omitted for a parameter that does not have a default value (which is a semantic error).

Comment (genlu) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I see, it'd be nice if the comment make this clear.

**Comment by genlu** [comment]
:+1:

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 3]
```diff
@@ -97,22 +97,23 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
...
                     {
+                        // The parameter is optional with a default value.
                         arguments.Add(new Argument(ArgumentKind.DefaultValue, parameter, new Literal(parameter.ExplicitDefaultConstantValue, parameter.Type, null)));
                     }
                     else
                     {
-                        // If corresponding parameter is Param array, then this means 0 element is provided and an Argument of kind == ParamArray will be added, 
-                        // otherwise it is an error and null is added.
-                        arguments.Add(DeriveArgument(parameterIndex, argumentIndex, boundArguments, argumentNames, argumentRefKinds, parameters, invocationSyntax));
+                        // If the invocation is semantically valid, the parameter will be a params array and an empty array will be provided.
```
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: At @genlu 's suggestion (which seems to have vanished), I've improved the comment.

**Comment by JohnHamby** [comment]
I'm going to merge into main now and work out how to merge into stabilization if the Ask Mode request is approved.

**Comment by mavasani** [comment]
@JohnHamby I verified that moving roslyn-analyzers to newer myget packages with your change (see this [PR](https://github.com/dotnet/roslyn-analyzers/pull/857)) and then re-running the repro steps doesn't hit this crash - so this change fixed the original repro.

Our Operation analyzers in roslyn-analyzers repo run fine on Roslyn.sln and Analyzers.sln (at least running for more than 30 mins and seen no crash/assert)

**Comment by Pilchie** [comment]
@mavasani It would be nice to include the string "Test Signoff" in your post above, since stashpop finds it in comments and extracts it.

File: src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb [hunk 4]
```diff
@@ -1191,6 +1191,15 @@ Class C
         M0(1, New Integer() { 2, 3, 4 })
         M0(1, New Integer() { 2, 3, 4, 5 })
         M0(1, New Integer() { 2, 3, 4, 5, 6 })
+        Dim local As D = new D(1, 2, 3, 4, 5)
```
Comment (Pilchie) [inline src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb]: Any reason not to include similar named parameter cases here in VB?

**Comment by Pilchie** [comment]
This LGTM, but I'll wait for the PR against `stabilization` to give the approval label.

File: src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb (hunk 4 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Test/Semantic/Diagnostics/OperationAnalyzerTests.vb]: Yes. VB prohibits writing the declaration of a method or constructor with a ParamArray after an Optional parameter, so constructing the test is difficult. The way the VB compiler processes argument lists prevents the issue from arising in VB anyway, so the specific issue cannot arise.

**Comment by genlu** [comment]
Here's the corresponding PR for stabilization: #8887

**Pull request merged**